jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/SaveDockerStep.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.builder.steps;
+
+import com.google.cloud.tools.jib.api.ImageReference;
+import com.google.cloud.tools.jib.configuration.BuildConfiguration;
+import com.google.cloud.tools.jib.docker.DockerClient;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.concurrent.Callable;
+
+/** Saves an image from the docker daemon. */
+public class SaveDockerStep implements Callable<Path> {
+
+  private final BuildConfiguration buildConfiguration;
+  private final DockerClient dockerClient;
+
+  SaveDockerStep(BuildConfiguration buildConfiguration, DockerClient dockerClient) {
+    this.buildConfiguration = buildConfiguration;
+    this.dockerClient = dockerClient;
+  }
+
+  @Override
+  public Path call() throws IOException, InterruptedException {
+    Path outputPath = Files.createTempDirectory("jib-docker-save").resolve("out.tar");
```

Comment:
Since the file will usually be huge, and we will create a new tar every time the user uses a local Docker image, I think it is good to try a cleanup later. (But don't crash if we fail to clean up.) Maybe we could explicitly delete (`MoreFiles.deleteRecursively()`) the directly once we know we'll no longer use it, or just use some built-in cleanup facilities. I remember something like `deleteOnExit()`. Not sure if it works for deleting files recursively. Or, we could use a [shutdown hook](https://github.com/GoogleCloudPlatform/appengine-plugins-core/blob/8316f066213315badc9a7353a428edc576ccb544/src/main/java/com/google/cloud/tools/managedcloudsdk/components/WindowsBundledPythonCopier.java#L58).

Maybe we should do this for the untarred directory? What do you think in general?

Reply:
I added shutdown hooks to delete both the saved tar and the extracted contents, although I think in the future we should probably cache the extracted layers instead of deleting them.